### ISSUE TITLE
The first term 哦分

### ISSUE THREAD

**Issue by PhyscalX** [zimenglan-sysu-512/Focal-Loss#2]

**Comment by zimenglan-sysu-512** [comment]
hi @PhyscalX,

here i just ignore the sign, because it can be dismissed by multiplying the ` -1 * p_i (p_i - 1)` or `-1 * p_j * p_i`, which is the derivative of normal cross-entropy loss using softmax function.  you can see the line `141` and `144` in .cu file.

**Comment by PhyscalX** [comment]
Your  mathematic consideration is right, the sign is dismissed.
But the "(power_prob_data[ind_i] / (1 - prob_data[ind_i])) "  multiplied by "(prob_data[ind_i] - 1)",
lead to "-power_prob_data[ind_i]" at the same time, actually it should be "power_prob_data[ind_i]"

**Comment by zimenglan-sysu-512** [comment]
hi @PhyscalX,

you are right, the loss is computed wrong, and thanks for reminding of `log` operation.  i will update my code tomorrow and do some test.

for the first term you mention, i need to double check.

thanks again.

**Comment by PhyscalX** [comment]
I have testified my idea on cifar10-quick, it is right，got the similar val acc as original loss @(alpha=1.0/0.75/0.5/0.25, gamma=2.0)

eps is very important in focal loss, all divisions in your code are dangerous, when alpha > 0.25, 
it tends to encounter NaN in cifar10-quick  at the end of convergence.

I preset eps as 1e-10 in my framework [Dragon](https://github.com/neopenx/Dragon).
If you are interested in it, check the following codes:

[(op_kernel.h, line 336), the declaration of kernel::SparseSoftmaxFocalLoss](https://github.com/neopenx/Dragon/blob/master/Dragon/include/utils/op_kernel.h)
[(op_kernel.cc, line 777), the CPU implementation of kernel::SparseSoftmaxFocalLoss](https://github.com/neopenx/Dragon/blob/master/Dragon/src/utils/op_kernel.cc)
[(op_kernel.cu, line 1417), the CUDA implementation of kernel::SparseSoftmaxFocalLoss](https://github.com/neopenx/Dragon/blob/master/Dragon/src/utils/op_kernel.cu)
[(sparse_softmax_focal_loss_op.h), the declaration of SparseSoftmaxFocalLossOp](https://github.com/neopenx/Dragon/blob/master/Dragon/include/operators/loss/sparse_softmax_focal_loss_op.h)
[(sparse_softmax_focal_loss_op.cc), the implementation of SparseSoftmaxFocalLossOp](https://github.com/neopenx/Dragon/blob/master/Dragon/src/operators/loss/sparse_softmax_focal_loss_op.cc)

**Comment by zimenglan-sysu-512** [comment]
hi @PhyscalX,

so thanks a lot. I have fixed the problems that u tell me. for the gradient, i forgot to derivate the (1 - p_t), so i ignore the sign.

thanks again.

**Comment by xudongyang** [comment]
hi @zimenglan-sysu-512 ,
I think there should be a minus sign in the grad of p_t w.r.t. x_j when i != j. Since you ignore the sign in the computation of grad of p_t, so when i=j, you change prob_data[ind_i] * (1 - prob_data[ind_i]) to prob_data[ind_i] * (prob_data[ind_i] - 1) in line 224. 
When i !=j, p_t = 1 - prob_data[ind_i], the grad of p_t w.r.t. x_j is prob_data[ind_i] * prob_data[ind_j]  then we should add back a minus sign right? Correct me if i am wrong

**Comment by zimenglan-sysu-512** [comment]
hi @PhyscalX,

i fix the `NaN` problem. i guess that, the op `(1 - p_t)^gamma / (1 - p_t)` causes the problem, when `1 - p_t` is so small.

now you can have a try.

**Comment by zimenglan-sysu-512** [comment]
hi @PhyscalX,

i don't think so, because it's multi-classes instead of binary classification. u can see the derivative of softmax to get more information.

**Comment by zimenglan-sysu-512** [comment]
hi @PhyscalX,

you're right, eps is very important, i add it to solve the `NaN` problem. Right now, it can run normally, you can have a look.

Thanks for pointing my error and giving so useful suggestions.

**Comment by PhyscalX** [comment]
Recommend you to multiply "grad" by prob_data[ind_i].
Directly dividing it may still lead to potential numerical issues.
Formulate in ONE way，and Implement in ANOTHER.
There are enormous tricks in programming mathematical formulations.

**Comment by zimenglan-sysu-512** [comment]
hi @PhyscalX,

i have updated. 
thanks a lot.
